# 1
Import the tools.

In [1]:
import pandas as pd
import numpy as np
import copy

# 2
Read the data.

In [2]:
flights = pd.read_csv('flights.csv')
sectorhours = pd.read_csv('sectorhours.csv')
trajectories = pd.read_csv('trajectories.csv')

# 3
Print the data.

In [3]:
print(flights)
print (sectorhours)
print (trajectories)

                                  uuid           O_D_A             dep_time  \
0      GRL780_BGSF_EKCH_20170901134000  BGSF_EKCH_A332  2017-09-01 13:50:00   
1      N174JF_BGSF_LOWI_20170901103500  BGSF_LOWI_AT43  2017-09-01 10:45:00   
2      WOW462_BIKF_EBBR_20170901065500  BIKF_EBBR_A321  2017-09-01 06:25:00   
3      ICE554_BIKF_EBBR_20170901074000  BIKF_EBBR_B752  2017-09-01 07:50:00   
4      WOW720_BIKF_EDDB_20170901054500  BIKF_EDDB_A321  2017-09-01 05:55:00   
...                                ...             ...                  ...   
29912   HVN31_VVTS_EDDF_20170901155500  VVTS_EDDF_A332  2017-09-01 16:05:00   
29913   HVN51_VVTS_EGLL_20170901173500  VVTS_EGLL_A332  2017-09-01 17:45:00   
29914  AFR253_VVTS_LFPG_20170901013000  VVTS_LFPG_A332  2017-09-01 01:40:00   
29915  THY166_VVTS_LTBA_20170901144000  VVTS_LTBA_A332  2017-09-01 14:50:00   
29916  BWJ998_ZYHB_EGGW_20170901110000  ZYHB_EGGW_E190  2017-09-01 11:10:00   

      original_trajectory  dep_time_minutes  
0    

# 4
Drop unnecessary columns in "flights".

In [4]:
flights = flights.drop(columns=['O_D_A', 'dep_time'])
flights

uuid original_trajectory  dep_time_minutes
0      GRL780_BGSF_EKCH_20170901134000    BGSF_EKCH_A332_3               830
1      N174JF_BGSF_LOWI_20170901103500    BGSF_LOWI_AT43_1               645
2      WOW462_BIKF_EBBR_20170901065500    BIKF_EBBR_A321_1               385
3      ICE554_BIKF_EBBR_20170901074000    BIKF_EBBR_B752_1               470
4      WOW720_BIKF_EDDB_20170901054500    BIKF_EDDB_A321_1               355
...                                ...                 ...               ...
29912   HVN31_VVTS_EDDF_20170901155500    VVTS_EDDF_A332_2               965
29913   HVN51_VVTS_EGLL_20170901173500    VVTS_EGLL_A332_3              1065
29914  AFR253_VVTS_LFPG_20170901013000    VVTS_LFPG_A332_3               100
29915  THY166_VVTS_LTBA_20170901144000    VVTS_LTBA_A332_1               890
29916  BWJ998_ZYHB_EGGW_20170901110000    ZYHB_EGGW_E190_1               670

[29917 rows x 3 columns]

# 5
Consider only the sectorhours with role 'E'. <br>Reset the index.

In [5]:
sectorhours = sectorhours[sectorhours['role']=='E']
sectorhours.reset_index(drop=True, inplace=True)
sectorhours

sector_hours      icao role  minute_start  minute_end  capacity
0       EGTTOCK_0000_0100_E   EGTTOCK    E             0          59        49
1       EGTTOCK_0100_0200_E   EGTTOCK    E            60         119        49
2       EGTTOCK_0200_0300_E   EGTTOCK    E           120         179        49
3       EGTTOCK_0300_0400_E   EGTTOCK    E           180         239        49
4       EGTTOCK_0400_0500_E   EGTTOCK    E           240         299        49
...                     ...       ...  ...           ...         ...       ...
13485   LSAGL67_2010_2040_E   LSAGL67    E          1210        1239        23
13486     LYBAL_1659_1800_E     LYBAL    E          1019        1079        41
13487    LYBATN_1600_1700_E    LYBATN    E           960        1019        43
13488    LYBAUN_1600_1700_E    LYBAUN    E           960        1019        43
13489  LYBAUWES_1659_1800_E  LYBAUWES    E          1019        1079        54

[13490 rows x 6 columns]

# 6
Merge "trajectories" with "flights".

E' indifferente applicare merge con modalità 'inner' o 'right': tutte le traiettorie in flights hanno un corrispondente in trajectories (il viceversa non è vero, come si potrebbe verificare applicando la modalità 'left': ci sono traiettorie in trajectories che non sono state volate il 1/9/2017).

In [6]:
merged = trajectories.merge(flights, how='right', left_on='centroid', right_on='original_trajectory')
merged

centroid                 eobt       icao           time_entry  \
0       EGKB_EGJB_OTHR_3  2017-09-07 15:00:00    EGTTFIS  2017-09-07 15:10:00   
1       EGKB_EGJB_OTHR_3  2017-09-07 15:00:00    EGTTTMS  2017-09-07 15:10:14   
2       EGKB_EGJB_OTHR_3  2017-09-07 15:00:00    EGTTBIG  2017-09-07 15:11:27   
3       EGKB_EGJB_OTHR_3  2017-09-07 15:00:00    EGTTTIM  2017-09-07 15:20:45   
4       EGKB_EGJB_OTHR_3  2017-09-07 15:00:00  EGTT18SFD  2017-09-07 15:27:54   
...                  ...                  ...        ...                  ...   
779662  LWSK_EDDF_AT43_1  2017-09-01 13:00:00    EDGGPSA  2017-09-01 14:52:59   
779663  LWSK_EDDF_AT43_1  2017-09-01 13:00:00    EDGGADF  2017-09-01 14:53:54   
779664  LMML_HLLB_OTHR_1  2017-02-16 21:45:00    LMMMWST  2017-02-16 22:11:47   
779665  LMML_HLLB_OTHR_1  2017-02-16 21:45:00    LMMMALL  2017-02-16 22:11:47   
779666  LMML_HLLB_OTHR_1  2017-02-16 21:45:00    HLLLALL  2017-02-16 22:39:10   

        time_to_sector_from_origin                             uuid  \
0                               10   N930Z_EGKB_EGJB_20170901101000   
1                               10   N930Z_EGKB_EGJB_20170901101000   
2                               11   N930Z_EGKB_EGJB_20170901101000   
3                               20   N930Z_EGKB_EGJB_20170901101000   
4                               27   N930Z_EGKB_EGJB_20170901101000   
...                            ...                              ...   
779662                         112   Z3MKD_LWSK_EDDF_20170901130000   
779663                         113   Z3MKD_LWSK_EDDF_20170901130000   
779664                          26  N907DB_LMML_HLLB_20170901170000   
779665                          26  N907DB_LMML_HLLB_20170901170000   
779666                          54  N907DB_LMML_HLLB_20170901170000   

       original_trajectory  dep_time_minutes  
0         EGKB_EGJB_OTHR_3               610  
1         EGKB_EGJB_OTHR_3               610  
2         EGKB_EGJB_OTHR_3               610  
3         EGKB_EGJB_OTHR_3               610  
4         EGKB_EGJB_OTHR_3               610  
...                    ...               ...  
779662    LWSK_EDDF_AT43_1               785  
779663    LWSK_EDDF_AT43_1               785  
779664    LMML_HLLB_OTHR_1              1030  
779665    LMML_HLLB_OTHR_1              1030  
779666    LMML_HLLB_OTHR_1              1030  

[779667 rows x 8 columns]

# 7
Drop unnecessary columns.
<br>Reorder the columns.
<br>Rename the columns.

In [7]:
merged = merged.drop(columns=['eobt', 'time_entry', 'original_trajectory'])
merged = merged[['uuid', 'centroid', 'icao', 'dep_time_minutes', 'time_to_sector_from_origin']]
merged.columns = ['flight', 'trajectory', 'icao', 'dep_time', 'time_to_sector']
merged

flight        trajectory       icao  \
0        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTFIS   
1        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTTMS   
2        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTBIG   
3        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTTIM   
4        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTT18SFD   
...                                 ...               ...        ...   
779662   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1    EDGGPSA   
779663   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1    EDGGADF   
779664  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    LMMMWST   
779665  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    LMMMALL   
779666  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    HLLLALL   

        dep_time  time_to_sector  
0            610              10  
1            610              10  
2            610              11  
3            610              20  
4            610              27  
...          ...             ...  
779662       785             112  
779663       785             113  
779664      1030              26  
779665      1030              26  
779666      1030              54  

[779667 rows x 5 columns]

# 8
Create a new column 'entry', describing the time a flight enters a sector-hour.
<br>The modulo operation is needed since we might exceed midnight (60x24=1440).

In [8]:
merged['entry'] = (merged['dep_time'] + merged['time_to_sector'])%1440
merged

flight        trajectory       icao  \
0        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTFIS   
1        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTTMS   
2        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTBIG   
3        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3    EGTTTIM   
4        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTT18SFD   
...                                 ...               ...        ...   
779662   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1    EDGGPSA   
779663   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1    EDGGADF   
779664  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    LMMMWST   
779665  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    LMMMALL   
779666  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1    HLLLALL   

        dep_time  time_to_sector  entry  
0            610              10    620  
1            610              10    620  
2            610              11    621  
3            610              20    630  
4            610              27    637  
...          ...             ...    ...  
779662       785             112    897  
779663       785             113    898  
779664      1030              26   1056  
779665      1030              26   1056  
779666      1030              54   1084  

[779667 rows x 6 columns]

# 9
Check that the maximum value of the column 'entry' doesn't exceed 1440.

In [9]:
print ("Is the maximum value of the column 'entry' smaller than 1440?")
print (max(merged['entry'])<1440)
print("max(merged['entry']) =", max(merged['entry']))

Is the maximum value of the column 'entry' smaller than 1440?
True
max(merged['entry']) = 1439


# 10
Merge "merged" with "sectorhours", based on 'icao'.

In [10]:
merged_2 = merged.merge(sectorhours, how='left', left_on='icao', right_on='icao')
merged_2

flight        trajectory     icao  dep_time  \
0         N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTTFIS       610   
1         N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTTTMS       610   
2         N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTTTMS       610   
3         N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTTTMS       610   
4         N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3  EGTTTMS       610   
...                                  ...               ...      ...       ...   
6841515  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1  LMMMALL      1030   
6841516  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1  LMMMALL      1030   
6841517  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1  LMMMALL      1030   
6841518  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1  LMMMALL      1030   
6841519  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1  HLLLALL      1030   

         time_to_sector  entry         sector_hours role  minute_start  \
0                    10    620                  NaN  NaN           NaN   
1                    10    620  EGTTTMS_0000_0100_E    E           0.0   
2                    10    620  EGTTTMS_0100_0200_E    E          60.0   
3                    10    620  EGTTTMS_0200_0300_E    E         120.0   
4                    10    620  EGTTTMS_0300_0400_E    E         180.0   
...                 ...    ...                  ...  ...           ...   
6841515              26   1056  LMMMALL_2000_2100_E    E        1200.0   
6841516              26   1056  LMMMALL_2100_2200_E    E        1260.0   
6841517              26   1056  LMMMALL_2200_2300_E    E        1320.0   
6841518              26   1056  LMMMALL_2300_2400_E    E        1380.0   
6841519              54   1084                  NaN  NaN           NaN   

         minute_end  capacity  
0               NaN       NaN  
1              59.0      40.0  
2             119.0      40.0  
3             179.0      40.0  
4             239.0      40.0  
...             ...       ...  
6841515      1259.0      39.0  
6841516      1319.0      39.0  
6841517      1379.0      39.0  
6841518      1439.0      39.0  
6841519         NaN       NaN  

[6841520 rows x 11 columns]

The modality 'left' keeps all the sectorhours present in "merged", even when they don't have any corrispondence in "sectorhours" (the relevant entries are set to 'Nan').
The modality 'inner', instead, would keep only the sectorhours in "merged" having an icao which is also present in "sectorhours", thus leaving out a number of flights.

# 11
Keep only the sectorhours having a 'minute_start' not greater than 'entry' and a 'minute_end' not smaller than 'entry'.
<br>Compare the number of flights included in "merged_2" and "clean2" respectively.

In [11]:
clean = merged_2[merged_2['minute_start']<=merged_2['entry']]
clean2 = clean[clean['entry']<=clean['minute_end']]

In [12]:
l1 = merged_2['flight'].unique()
l2 = clean2['flight'].unique()
print("len(l1): ", len(l1))
print("len(l2): ", len(l2))
diff = list(set(l1) - set(l2))
print("len(diff): ", len(diff))

len(l1):  29917
len(l2):  29858
len(diff):  59


The number of flights being equal to 29858, tells us that 59 flights are left out.
This is due to the fact that the merge modality 'left' keeps all the sectorhours present in "merged", even when they don't have any corrispondence in "sectorhours"; the missing entries are then set to 'NaN', but the inequalities of step 11 leave out the rows corresponding to such entries.
The 59 missing flights are included again at a later step (#14 and 15).

# 12
Drop unnecessary columns. <br>Reset the index.

In [14]:
clean2 = clean2.drop(columns=['icao', 'time_to_sector', 'entry', 'role', 'minute_start', 'minute_end', 'capacity'])
clean2.reset_index(drop=True, inplace=True)
clean2

flight        trajectory  dep_time  \
0        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3       610   
1        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3       610   
2        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3       610   
3        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3       610   
4        N930Z_EGKB_EGJB_20170901101000  EGKB_EGJB_OTHR_3       610   
...                                 ...               ...       ...   
264896   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1       785   
264897   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1       785   
264898   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1       785   
264899   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43_1       785   
264900  N907DB_LMML_HLLB_20170901170000  LMML_HLLB_OTHR_1      1030   

                 sector_hours  
0         EGTTTMS_1000_1100_E  
1         EGTTBIG_1000_1100_E  
2         EGTTTIM_1000_1100_E  
3       EGTT18SFD_1000_1100_E  
4       EGTT19HRE_1000_1100_E  
...                       ...  
264896  EDUUDON1D_1400_1500_E  
264897    EDMMALB_1400_1500_E  
264898    EDGGKTG_1400_1500_E  
264899    EDGGPSA_1400_1500_E  
264900    LMMMALL_1731_1800_E  

[264901 rows x 4 columns]

# 13
Collapse all the sectorhours associated to each flight into a list of sectorhours.
<br>A new dataframe is created, containing the two columns 'flight' and 'sector_hours'.

In [15]:
clean3 = clean2.groupby('flight')
clean4 = clean3['sector_hours'].unique()

clean5 = pd.DataFrame({'flight':clean4.index, 'sector_hours':clean4.values})

clean5

flight  \
0        A6RTS_OMDW_EGGW_20170901055500   
1       AAB249_EBKT_LFMN_20170901070500   
2       AAB249_LFMN_LIEO_20170901090500   
3       AAB25C_LOWW_UUMO_20170901143500   
4       AAB571_EBBR_EGNV_20170901050500   
...                                 ...   
29853   XLF056_LFPG_MMUN_20170901153100   
29854  XLF1451_LFML_LFPG_20170901030500   
29855   XRO546_LFMD_EGCC_20170901180000   
29856   XRO552_EHAM_LFMD_20170901150000   
29857    Z3MKD_LWSK_EDDF_20170901130000   

                                            sector_hours  
0      [LTAAAEU67_0800_0900_E, LTAAANOR2_0900_1000_E,...  
1      [EBBUWLS_0700_0800_E, EBBUWHS_0700_0800_E, EDY...  
2      [LFMMMNST_0900_1000_E, LFMMBTAJ_0900_1000_E, L...  
3      [LZBBL33_1500_1600_E, LZBBU_1500_1600_E, LZBBU...  
4      [EBBUNWC_0530_0600_E, EDYYB5KL_0500_0600_E, ED...  
...                                                  ...  
29853  [LFFFTH_1600_1700_E, LFFFUZ_1600_1700_E, LFRRJ...  
29854  [LFMMMALY_0330_0400_E, LFMMWW_0330_0400_E, LFM...  
29855  [LFMMSBAM_1800_1900_E, LFMMB12_1800_1900_E, LF...  
29856  [EHAACOD_1500_1600_E, EHAASECT3_1520_1600_E, E...  
29857  [LWSSLOW_1300_1400_E, LYBAL_1300_1359_E, LQSBU...  

[29858 rows x 2 columns]

# 14
Recover the 59 missing flights: create a new dataframe "flights2" selecting from "flights" only the rows having an 'uuid' that is present in the list "diff". <br>Drop the unnecessary columns. <br>Rename the column; add a new column named 'sector_hours', with empty lists as entries.
<br>Check that the lenght of this new dataframe is 59.

In [16]:
flights2 = flights[flights['uuid'].isin(diff)]
flights2 = flights2.drop(columns=['original_trajectory', 'dep_time_minutes'])
flights2.columns = ['flight']
flights2['sector_hours'] = [[] for _ in range(len(flights2))]
    
print(len(flights2))    
flights2

59


flight sector_hours
4768     AOS54_EGBJ_EGFF_20170901092000           []
5141    EZE612_EGFF_EGOV_20170901142500           []
5142    EZE610_EGFF_EGOV_20170901063500           []
5540     EGL08_EGHH_EGTE_20170901181500           []
7004     MDI09_EGNT_EGNV_20170901191500           []
7209    EZE613_EGOV_EGFF_20170901154000           []
7210    EZE611_EGOV_EGFF_20170901075500           []
7874   NJE347D_EGSS_EGGW_20170901193000           []
8897    CAI040_EHEH_EHBK_20170901074000           []
8948    HNL33A_EHFE_EHKD_20170901104500           []
8949    HNL63C_EHFQ_EHKD_20170901163000           []
8950    HNL25A_EHFT_EHKD_20170901080000           []
8960    HNL05B_EHKP_EHKD_20170901170000           []
8961    HNL70A_EHLF_EHKD_20170901101500           []
8962    HNL07A_EHLF_EHKD_20170901074500           []
8963    HNL07C_EHLF_EHKD_20170901170000           []
8964    HNL05A_EHLL_EHKD_20170901114000           []
8965    NHX15B_EHLT_EHKD_20170901173000           []
8966    HHE60B_EHLX_EHKD_20170901130900           []
8967    NHX01A_EHMF_EHKD_20170901064000           []
8968    HNL63A_EHMG_EHKD_20170901063500           []
8969    HNL25C_EHMR_EHKD_20170901163000           []
8970    HNL54A_EHPG_EHKD_20170901062000           []
8971    HNL24B_EHPM_EHKD_20170901174500           []
8972    NHX01B_EHQH_EHKD_20170901160000           []
8973    NHX15A_EHQI_EHKD_20170901073000           []
14454   AOJ75D_LDVA_LQBK_20170901053000           []
17405   NAK747_LFGJ_LFLA_20170901103000           []
17724  TJT052P_LFMD_LFML_20170901120000           []
17802  HOP15VJ_LFML_LFBO_20170901143000           []
17803  HOP17VJ_LFML_LFBO_20170901192500           []
17826   FPO759_LFML_LFMN_20170901004000           []
18008  FPO750F_LFMN_LFML_20170901190500           []
18009   RIV202_LFMN_LFML_20170901180000           []
18215   NAK401_LFOK_LFPM_20170901071500           []
18914   NAK401_LFPM_LFOK_20170901063000           []
19245    N166G_LFPT_LFPN_20170901092000           []
19514  TVS2456_LFSB_LFST_20170901110000           []
19535  GAC018Z_LFSB_LSZH_20170901081000           []
23949  AUA165Q_LOWG_LOWI_20170901200500           []
24288  AUA935T_LOWW_LOWK_20170901075500           []
24289   AUA99J_LOWW_LOWK_20170901151500           []
24290   AUA93J_LOWW_LOWK_20170901210500           []
24291  AUA929K_LOWW_LOWK_20170901181500           []
24293  GDK01CA_LOWW_LOWK_20170901100000           []
28721  FDB8503_OMDB_UGTB_20170901013000           []
28722   FDB713_OMDB_UGTB_20170901213000           []
28912  AHY9223_UBBB_UGTB_20170901043000           []
29026   TGZ785_UGTB_ULLI_20170901172000           []
29032  AFL1891_UGTB_UUEE_20170901092000           []
29033  AFL1893_UGTB_UUEE_20170901231500           []
29319   FIN166_ULLI_EFHK_20170901090000           []
29320   FIN168_ULLI_EFHK_20170901024500           []
29321   FIN170_ULLI_EFHK_20170901164000           []
29383   TGZ786_ULLI_UGTB_20170901214500           []
29614   SBI955_UUDD_UGTB_20170901105500           []
29647  AFL2200_UUEE_EFHK_20170901152000           []
29648  AFL2206_UUEE_EFHK_20170901074500           []
29649   CLU517_UUEE_EFHK_20170901132000           []

# 15
Append "flights2" to "clean".
<br>Sort the complete dataframe accoridng to the column 'flight'.
<br>Reset the index.

In [17]:
clean5 = clean5.append(flights2)
clean5.sort_values(by=['flight'], inplace=True)
clean5.reset_index(drop=True, inplace=True)
clean5

flight  \
0        A6RTS_OMDW_EGGW_20170901055500   
1       AAB249_EBKT_LFMN_20170901070500   
2       AAB249_LFMN_LIEO_20170901090500   
3       AAB25C_LOWW_UUMO_20170901143500   
4       AAB571_EBBR_EGNV_20170901050500   
...                                 ...   
29912   XLF056_LFPG_MMUN_20170901153100   
29913  XLF1451_LFML_LFPG_20170901030500   
29914   XRO546_LFMD_EGCC_20170901180000   
29915   XRO552_EHAM_LFMD_20170901150000   
29916    Z3MKD_LWSK_EDDF_20170901130000   

                                            sector_hours  
0      [LTAAAEU67_0800_0900_E, LTAAANOR2_0900_1000_E,...  
1      [EBBUWLS_0700_0800_E, EBBUWHS_0700_0800_E, EDY...  
2      [LFMMMNST_0900_1000_E, LFMMBTAJ_0900_1000_E, L...  
3      [LZBBL33_1500_1600_E, LZBBU_1500_1600_E, LZBBU...  
4      [EBBUNWC_0530_0600_E, EDYYB5KL_0500_0600_E, ED...  
...                                                  ...  
29912  [LFFFTH_1600_1700_E, LFFFUZ_1600_1700_E, LFRRJ...  
29913  [LFMMMALY_0330_0400_E, LFMMWW_0330_0400_E, LFM...  
29914  [LFMMSBAM_1800_1900_E, LFMMB12_1800_1900_E, LF...  
29915  [EHAACOD_1500_1600_E, EHAASECT3_1520_1600_E, E...  
29916  [LWSSLOW_1300_1400_E, LYBAL_1300_1359_E, LQSBU...  

[29917 rows x 2 columns]

# 16
Export the dataframe as pickle (in order to preserve the type of the objects).

In [18]:
clean5.to_pickle('flights_sectorhours.pkl')